### PR TITLE
Use page title for filename with :download

### PR DESCRIPTION
**Pull request by iordanisg**
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/2753)

### CODE REVIEW

**Comment by iordanisg** [comment]
Per #2652

File: qutebrowser/browser/commands.py [hunk 1]
```diff
@@ -1444,7 +1444,8 @@ def download(self, url=None, dest_old=None, *, mhtml_=False, dest=None):
         else:
             qnam = tab.networkaccessmanager()
             download_manager.get(self._current_url(), user_agent=user_agent,
-                                 qnam=qnam, target=target)
+                                 qnam=qnam, target=target,
+                                 title=self._current_title())
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: I think passing a title here makes `get_request` needlessly complicated for only this single use case.

Instead, you should be able to construct a `downloads.FileDownloadTarget` here (similar to what's done above when `--dest` is given), and pass that to `.get()` instead. Then there shouldn't be any changes in `qtnetworkdownloads.py` needed at all.

File: qutebrowser/browser/qtnetworkdownloads.py [hunk 2]
```diff
@@ -429,7 +429,8 @@ def get_request(self, request, *, target=None, **kwargs):
                              QNetworkRequest.AlwaysNetwork)
 
         if request.url().scheme().lower() != 'data':
-            suggested_fn = urlutils.filename_from_url(request.url())
+            suggested_fn = (utils.sanitize_filename(title) + ".html" if title
```
Comment (The-Compiler) [inline qutebrowser/browser/qtnetworkdownloads.py]: Oh, good job at finding `utils.sanitize_filename`! I compltely forgot about it 😆

**Comment by Kingdread** [comment]
I think this method is a bit too simplistic, as it always takes the page title, even for non-HTML pages:

```
[daniel@persephone]  ~ % ls /tmp/qutebrowser-basedir-4mql1yi0/download/
'Vanamo_Logo.png (1200x1200 pixels).html'
[daniel@persephone]  ~ % file /tmp/qutebrowser-basedir-4mql1yi0/download/Vanamo_Logo.png\ \(1200x1200\ pixels\).html
/tmp/qutebrowser-basedir-4mql1yi0/download/Vanamo_Logo.png (1200x1200 pixels).html: PNG image data, 1200 x 1200, 4-bit colormap, non-interlaced
```

(the old method produced a `Vanamo_Logo.png` file, which is fine)

I think the question is: How do we determine if we got a sensible filename out of the URL, or when should we use the page title?

**Comment by The-Compiler** [comment]
Hmm, damn. This is indeed something I didn't consider, and a good point...

I have two ideas:

- Use the title when there's no path in the URL, and have a whitelist of file extensions where it makes sense to do so too (as a first approximation, `.html`, `.htm`, `.php`, but probably more)
- Don't do this at all. 😟

**Comment by Kingdread** [comment]
I'd probably add "if no extension", as many modern pages don't even use extensions anymore (for example, GitHub: `https://github.com/qutebrowser/qutebrowser/pull/2753`).

Alternatively, we could put this behind a flag (`:download --title`?), but then you would need to remember to use the flag.

**Comment by The-Compiler** [comment]
Any update, @iordanisg?

**Comment by The-Compiler** [comment]
Whoops, didn't mean to close that one, sorry.

**Comment by iordanisg** [comment]
@The-Compiler: your whitelist idea sounds good to me, I hope I'll be able to work something out soon (busy week so far)

@Kingdread: thanks for the feedback!

**Comment by iordanisg** [comment]
About the failing download test: a page title is created on the fly (even if the page doesn't have one) when running the tests.

File: qutebrowser/browser/commands.py [hunk 3]
```diff
@@ -1443,8 +1443,23 @@ def download(self, url=None, dest_old=None, *, mhtml_=False, dest=None):
...
             qnam = tab.networkaccessmanager()
-            download_manager.get(self._current_url(), user_agent=user_agent,
-                                 qnam=qnam, target=target)
+
+            # Downloads of URLs with file extensions in the whitelist will use
+            # the page title as the filename.
+            ext_whitelist = [".html", ".htm", ".php", ""]
+            _, ext = os.path.splitext(self._current_url().path())
+            if ext.lower() in ext_whitelist and tab.title():
+                suggested_fn = utils.sanitize_filename(tab.title()) + ext
```
Comment (The-Compiler) [inline qutebrowser/browser/commands.py]: When QtWebEngine auto-generates a title, this ends up as `no title.html.html`. I'd suggest checking for `if not tab.title().endswith(ext):` before adding it.

File: tests/end2end/features/downloads.feature [hunk 4]
```diff
@@ -22,6 +22,27 @@ Feature: Downloading things from a website.
...
+        Then the downloaded file Simple downloads.html should exist
+
+    Scenario: Using :download with no URL on an image
+        When I set storage -> prompt-download-directory to false
+        And I open data/downloads/qutebrowser.png
+        And I run :download
+        And I wait until the download is finished
+        Then the downloaded file qutebrowser.png should exist
+
+    Scenario: Using :download with no URL and no page title
```
Comment (The-Compiler) [inline tests/end2end/features/downloads.feature]: It looks like QtWebEngine autogenerates a filename while QtWebKit doesn't, which also explains that you don't see it locally. After adding some unittests (see my review comment), I'd suggest just removing this end2end test entirely (but keep the other two).

File: qutebrowser/browser/downloads.py [hunk 5]
```diff
@@ -182,6 +182,26 @@ def transform_path(path):
     return path
 
 
+def suggested_fn_from_title(url, title=None):
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: `url` should be called `url_path` to make it clear that it isn't a full URL.

File: qutebrowser/browser/downloads.py [hunk 6]
```diff
@@ -182,6 +182,26 @@ def transform_path(path):
...
+    """Suggest a filename depending on the URL extension and page title.
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: nitpick: Add an empty line here

File: qutebrowser/browser/downloads.py [hunk 7]
```diff
@@ -182,6 +182,26 @@ def transform_path(path):
...
+    Args:
+        url: a string with the URL path
+        title: the page title string
+
+    Returns None if the extension is not in the whitelist
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: nitpick: Put this under a `Return:` similar like with `Args:`

File: tests/unit/browser/webkit/test_downloads.py [hunk 8]
```diff
@@ -30,6 +30,30 @@ def test_download_model(qapp, qtmodeltester, config_stub, cookiejar_and_cache):
...
+@pytest.mark.parametrize('url, title, out', [
+    ('https://qutebrowser.org/img/cheatsheet-big.png',
+     'cheatsheet-big.png (3342×2060)',
+     None),
+    ('http://qutebrowser.org/INSTALL.html',
+     'Installing qutebrowser | qutebrowser',
+     'Installing qutebrowser _ qutebrowser.html'),
+    ('http://qutebrowser.org/INSTALL.html.html',
+     'Installing qutebrowser | qutebrowser',
+     'Installing qutebrowser _ qutebrowser.html'),
```
Comment (The-Compiler) [inline tests/unit/browser/webkit/test_downloads.py]: I don't understand this test - I think to resemble the behavior with the double extension, this should have a `INSTALL.html` filename (not `.html.html`) and a title which ends with `.html`, no?

File: qutebrowser/browser/downloads.py [hunk 9]
```diff
@@ -182,6 +182,26 @@ def transform_path(path):
...
+    or if there is no page title.
+    """
+    ext_whitelist = [".html", ".htm", ".php", ""]
+    _, ext = os.path.splitext(url)
+    if ext.lower() in ext_whitelist and title:
+        suggested_fn = utils.sanitize_filename(title)
+        if not suggested_fn.endswith(ext):
```
Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: Maybe `if not suggested_fn.lower().endswith(ext.lower())`?

File: tests/unit/browser/webkit/test_downloads.py [hunk 10]
```diff
@@ -30,6 +30,30 @@ def test_download_model(qapp, qtmodeltester, config_stub, cookiejar_and_cache):
...
+    ('http://qutebrowser.org/',
+     'qutebrowser | qutebrowser',
+     'qutebrowser _ qutebrowser'),
+    ('https://github.com/qutebrowser/qutebrowser/releases',
+     'Releases · qutebrowser/qutebrowser',
+     'Releases · qutebrowser_qutebrowser'),
+    ('http://qutebrowser.org/page-with-no-title.html',
+     '',
+     None),
+])
```
Comment (The-Compiler) [inline tests/unit/browser/webkit/test_downloads.py]: Please also add a test with an upper-case extension (to make sure the title is used), and with an upper-case `.HTML` in the title (see comment above).

**Comment by The-Compiler** [comment]
Looks good to me now - @Kingdread what do you think?

File: qutebrowser/browser/downloads.py [hunk 11]
```diff
@@ -182,6 +182,28 @@ def transform_path(path):
...
+
+    Return:
+        The download filename based on the title, or None if the extension is
+        not found in the whitelist (or if there is no page title).
+    """
+    ext_whitelist = [".html", ".htm", ".php", ""]
+    _, ext = os.path.splitext(url_path)
+    if ext.lower() in ext_whitelist and title:
+        suggested_fn = utils.sanitize_filename(title)
+        if not suggested_fn.lower().endswith(ext.lower()):
```
Comment (Kingdread) [inline qutebrowser/browser/downloads.py]: From the test cases and the code I assume that the `.html` is not appended if it's not in the original URL? Is that intended? I'd assume a title like "qutebrowser home page" on "http://qutebrowser.org" would result in a file `qutebrowser home page.html`.

Comment (Kingdread) [inline qutebrowser/browser/downloads.py]: Also, won't this add `.php` even for HTML files? Like you get a `index.php` locally, even if it's just the produced HTML content?

Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: Hmm, good points - maybe it's be best to check if it ends in either `.htm` or `.html` and if not, just append `.html`?

Comment (iordanisg) [inline qutebrowser/browser/downloads.py]: Very good points, indeed. What about when there is an uppercase `.HTML` or `.HTM` in the title? Would it make sense to keep it that way in the filename or convert it to lower case?

Comment (The-Compiler) [inline qutebrowser/browser/downloads.py]: I think it'd be fine to keep it, but I don't have a strong opinion either way. I'd go for "whatever is easier to implement" for this one.

**Comment by The-Compiler** [comment]
@Kingdread anything else? Otherwise, this looks ready to merge to me.

**Comment by Kingdread** [comment]
No, no more comments 😛

**Comment by The-Compiler** [comment]
And merged! Thanks @Kingdread for the reviews, and @iordanisg for the contribution and your patience! :smile:

**Comment by iordanisg** [comment]
Thank you both!

**Pull request merged**